### PR TITLE
Add more globalnet metrics

### LINKED ISSUE #1471: Support metrics in Globalnet 2.0
**What would you like to be added**:
In Globalnet 1.0, we used to have the following metrics. This issue is to support metrics with Globalnet 2.0 code introducing new metrics as necessary.

```
submariner_global_IP_availability | cidr | Count of available global IPs per CIDR
submariner_global_IP_allocated   | cidr | Count of global IPs allocated for Pods/Services per CIDR
```

### PR DESCRIPTION
**Pull request by maayanf24**
Signed-off-by: Maayan Friedman <maafried@redhat.com>

### CODE REVIEW

File: pkg/globalnet/controllers/base_controllers.go [hunk 1]
```diff
@@ -120,7 +120,9 @@ func (c *baseIPAllocationController) reserveAllocatedIPs(federator federate.Fede
 	err := c.pool.Reserve(reservedIPs...)
 
 	if err == nil && len(reservedIPs) > 0 {
+		metricsAllocate(c.pool.GetCider(), len(reservedIPs))
```
Comment (sridhargaddam) [inline pkg/globalnet/controllers/base_controllers.go]: We should move this to Line 130 as there is a possibility for error during postReserve and the allocated IP can be released in Line127

File: pkg/globalnet/controllers/base_controllers.go [hunk 2]
```diff
@@ -96,7 +96,7 @@ func (c *baseSyncerController) reconcile(client dynamic.ResourceInterface, label
 }
 
 func (c *baseIPAllocationController) reserveAllocatedIPs(federator federate.Federator, obj *unstructured.Unstructured,
-	postReserve func(allocatedIPs []string) error) error {
+	postReserve func(allocatedIPs []string) error, metricsAllocate func(string, int)) error {
```
Comment (sridhargaddam) [inline pkg/globalnet/controllers/base_controllers.go]: @tpantelis are you fine with this approach or do you have any suggestion?

File: pkg/globalnet/metrics/metrics.go [hunk 3]
```diff
@@ -43,10 +43,38 @@ var (
 			cidrLabel,
 		},
 	)
+	globalEgressIPsAllocatedGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "submariner_global_Egress_IP_allocated",
```
Comment (tpantelis) [inline pkg/globalnet/metrics/metrics.go]:
```suggestion
			Name: "submariner_global_egress_IP_allocated",
```

File: pkg/globalnet/metrics/metrics.go [hunk 4]
```diff
@@ -43,10 +43,38 @@ var (
...
+			Help: "Count of global Egress IPs allocated for Pods/Services per CIDR",
+		},
+		[]string{
+			cidrLabel,
+		},
+	)
+	clusterGlobalEgressIPsAllocatedGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "submariner_cluster_global_Egress_IP_allocated",
```
Comment (tpantelis) [inline pkg/globalnet/metrics/metrics.go]:
```suggestion
			Name: "submariner_cluster_global_egress_IP_allocated",
```

File: pkg/globalnet/metrics/metrics.go [hunk 5]
```diff
@@ -43,10 +43,38 @@ var (
...
+			Help: "Count of global Egress IPs allocated for clusters per CIDR",
+		},
+		[]string{
+			cidrLabel,
+		},
+	)
+	globalIngressIPsAllocatedGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "submariner_global_Ingress_IP_allocated",
```
Comment (tpantelis) [inline pkg/globalnet/metrics/metrics.go]:
```suggestion
			Name: "submariner_global_ingress_IP_allocated",
```

File: pkg/ipam/ippool.go [hunk 6]
```diff
@@ -214,3 +214,7 @@ func (p *IPPool) Size() int {
 
 	return p.available.Size()
 }
+
+func (p *IPPool) GetCider() string {
```
Comment (tpantelis) [inline pkg/ipam/ippool.go]:
```suggestion
func (p *IPPool) GetCIDR() string {
```

File: pkg/globalnet/controllers/base_controllers.go (hunk 2 above)
Comment (tpantelis) [reply pkg/globalnet/controllers/base_controllers.go]: I assume you mean adding a second callback function. I think it's simpler to also use the `postReserve` function for the metrics.

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Comment by sridhargaddam** [comment]
@maayanf24 can you please look into the review comments from @tpantelis and also rebase the PR.

File: pkg/globalnet/controllers/global_ingressip_controller.go [hunk 7]
```diff
@@ -310,7 +319,7 @@ func (c *globalIngressIPController) onDelete(ingressIP *submarinerv1.GlobalIngre
 		}
 
 		return nil
-	}, ingressIP.Status.AllocatedIP)
+	})
```
Comment (maayanf24) [inline pkg/globalnet/controllers/global_ingressip_controller.go]: this line removed by mistake, I'll add it back

File: pkg/globalnet/controllers/cluster_egressip_controller.go [hunk 8]
```diff
@@ -254,6 +259,8 @@ func (c *clusterGlobalEgressIPController) allocateGlobalIPs(key string, numberOf
 		return true
 	}
 
+	metrics.RecordAllocateClusterGlobalEgressIPs(c.pool.GetCIDR(), numberOfIPs)
```
Comment (tpantelis) [inline pkg/globalnet/controllers/cluster_egressip_controller.go]: I think we should move this after L279 then we don't need to record deallocated on L276 (it's a noop).

File: pkg/globalnet/controllers/global_egressip_controller.go [hunk 9]
```diff
@@ -81,6 +82,7 @@ func NewGlobalEgressIPController(config *syncer.ResourceSyncerConfig, pool *ipam
 
 	for i := range list.Items {
 		err = controller.reserveAllocatedIPs(federator, &list.Items[i], func(reservedIPs []string) error {
+			metrics.RecordAllocateGlobalEgressIPs(pool.GetCIDR(), len(reservedIPs))
```
Comment (tpantelis) [inline pkg/globalnet/controllers/global_egressip_controller.go:85]: This is on startup to reserve previously allocated IPs so it I don't think we should record this as we're not actually allocating new block. @sridhargaddam WDYT?

File: pkg/globalnet/controllers/global_egressip_controller.go [hunk 10]
```diff
@@ -226,6 +228,8 @@ func (c *globalEgressIPController) allocateGlobalIPs(key string, numberOfIPs int
 		return true
 	}
 
+	metrics.RecordAllocateGlobalEgressIPs(c.pool.GetCIDR(), numberOfIPs)
```
Comment (tpantelis) [inline pkg/globalnet/controllers/global_egressip_controller.go]: Similar to above, I think we should move this after L249 and remove L245.

File: pkg/globalnet/controllers/global_ingressip_controller.go [hunk 11]
```diff
@@ -81,6 +82,7 @@ func NewGlobalIngressIPController(config *syncer.ResourceSyncerConfig, pool *ipa
 
 		// nolint:wrapcheck  // No need to wrap these errors.
 		err = controller.reserveAllocatedIPs(federator, obj, func(reservedIPs []string) error {
+			metrics.RecordAllocateGlobalIngressIPs(pool.GetCIDR(), len(reservedIPs))
```
Comment (tpantelis) [inline pkg/globalnet/controllers/global_ingressip_controller.go:85]: Similar as above - we're reserving not allocating here.

File: pkg/globalnet/controllers/global_ingressip_controller.go [hunk 12]
```diff
@@ -163,12 +165,15 @@ func (c *globalIngressIPController) onCreate(ingressIP *submarinerv1.GlobalIngre
 		return true
 	}
 
+	metrics.RecordAllocateGlobalIngressIPs(c.pool.GetCIDR(), 1)
```
Comment (tpantelis) [inline pkg/globalnet/controllers/global_ingressip_controller.go]: Same as above (move after L255 and don't deallocate on error).

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/globalnet/controllers/global_egressip_controller.go (hunk 9 above)
Comment (sridhargaddam) [reply pkg/globalnet/controllers/global_egressip_controller.go:85]: Good point. 

1. If the metrics value is like a time series information on when the globalIPs are allocated then, as you said, we are just reserving the previously allocated globalIP here, so its not like a new allocation.
2. But if it shows the current number of globalIPs allocated, then we can record it (as the pod was just restarted).

 Lets hear from @skitt

Comment (skitt) [reply pkg/globalnet/controllers/global_egressip_controller.go:85]: The metric is maintained per run of the controller: whenever the controller starts, the metric is 0. If there are already-allocated IPs, we need to adjust the metric accordingly, so this seems correct to me.

**Comment by nyechiel** [comment]
@sridhargaddam @skitt it looks like Tom's comments have been addressed. Can you please review?

**Review by tpantelis — APPROVED** [review]

**Review by skitt — APPROVED** [review]

**Comment by sridhargaddam** [comment]
@skitt I remember you having some plans to have some CI job which checks for commit messages that includes message like "Address review comments", "Fix review comments" etc. The commits in this PR should have been squashed.

**Comment by skitt** [comment]
> @skitt I remember you having some plans to have some CI job which checks for commit messages that includes message like "Address review comments", "Fix review comments" etc. The commits in this PR should have been squashed.

Aargh, they should have, I missed that. The check already exists, it’s “No "Apply suggestions from code review" Commits”, but it only checks for GitHub suggestions and fix-ups.

**Pull request merged**